File: binary-search/C_Cpp/search_2d_matrix.c
#include <stdbool.h>

bool searchMatrix(int** matrix, int matrixSize, int* matrixColSize, int target) {
    int low = 0;
    int high = matrixSize * (*matrixColSize) - 1;

    while (low <= high) {
        int mid = low + (high - low) / 2;  // Mid index in the "virtual" 1D array

        // Convert mid index to 2D coordinates
        int midRow = mid / (*matrixColSize);   // Row index
        int midCol = mid % (*matrixColSize);   // Column index

        if (matrix[midRow][midCol] == target) {
            return true;
        } else if (matrix[midRow][midCol] < target) {
            low = mid + 1;  // Search in the right half
        } else {
            high = mid - 1;  // Search in the left half
        }
    }

    return false;  // Target not found
}

File: binary-search/Python/search_2d_matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        f = []
        for row in matrix:
            for elem in row:
                f.append(elem)
        low, high = 0, len(f) - 1
        while (low <= high):
            m = (low + high) // 2
            if target < f[m]:
                high = m - 1
            elif target > f[m]:
                low = m + 1
            else:
                return True
        return False

